app/src/main/java/de/hhn/aib/labsw/blackmirror/controller/API/Websockets/MirrorApiWebsockets.kt

```diff
@@ -0,0 +1,118 @@
+package de.hhn.aib.labsw.blackmirror.controller.API.Websockets
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.ObjectMapper
+import okhttp3.*
+import java.io.IOException
+
+class MirrorApiWebsockets : WebSocketListener(),MirrorApi {
+    var sessions: ArrayList<WebSocket> = ArrayList()
+    private var mapper: ObjectMapper = ObjectMapper()
+    var listeners: HashMap<String, ArrayList<ApiListener>?> =
+        HashMap()
+
+    fun getJSONMapper(): ObjectMapper {
+        return mapper
+    }
+
+    override fun onOpen(webSocket: WebSocket, response: Response) {
+        sessions.add(webSocket)
+        print("new Connection!: ")
+
+        /*Location l = new Location();
+        l.lat = 3.123;
+        l.lon = 1.452;
+        publish("location", l);*/
+    }
+
+    override fun onMessage(webSocket: WebSocket, text: String) {
+        try {
+            val jsonNode: JsonNode = mapper.readTree(text)
+            val topic: String = jsonNode.get("topic").textValue()
+            requireNotNull(jsonNode.get("payload")) { "wrong json format" }
+            val listenersList: ArrayList<ApiListener>? = listeners[topic]
+            listenersList?.forEach { element : ApiListener ->
+                try {
+                    element.dataReceived(topic, jsonNode)
+                } catch (e: NullPointerException) {
+                    listenersList.remove(element)
```

Comment:
wouldnt this throw a ConcurrentModificationException? u are usually not allowed to remove elements form a collection while iterating over it as this would invalidate the iterators